File: PramaIA-LogService/clients/javascript/pramaialog.js
/**
 * Client JavaScript per il servizio di logging PramaIA.
 * 
 * Questo modulo fornisce un client per inviare log al servizio
 * PramaIA-LogService. Può essere utilizzato dai componenti JavaScript/TypeScript
 * dell'ecosistema PramaIA, come il PDK.
 * 
 * @example
 * ```javascript
 * const { PramaIALogger, LogLevel, LogProject } = require('pramaialog');
 * 
 * // Crea un'istanza del logger
 * const logger = new PramaIALogger({
 *   apiKey: 'your_api_key',
 *   project: LogProject.PDK,
 *   module: 'workflow_editor'
 * });
 * 
 * // Invia log di diversi livelli
 * logger.info('Editor avviato');
 * logger.warning('Attenzione: configurazione nodo non valida', 
 *               { nodeId: 'node-123', issues: ['Missing input'] });
 * logger.error('Errore durante il salvataggio del workflow', 
 *             { workflowId: '123', error: err.message },
 *             { userId: 'admin' });
 * ```
 */

/**
 * Livelli di log supportati
 * @enum {string}
 */
const LogLevel = {
  DEBUG: 'debug',
  INFO: 'info',
  WARNING: 'warning',
  ERROR: 'error',
  CRITICAL: 'critical'
};

/**
 * Progetti PramaIA supportati
 * @enum {string}
 */
const LogProject = {
  SERVER: 'PramaIAServer',
  PDK: 'PramaIA-PDK',
  AGENTS: 'PramaIA-Agents',
  PLUGINS: 'PramaIA-Plugins',
  OTHER: 'other'
};

/**
 * Client per il servizio di logging PramaIA.
 */
class PramaIALogger {
  /**
   * Crea un'istanza del logger PramaIA.
   * 
   * @param {Object} options - Opzioni di configurazione
   * @param {string} options.apiKey - API key per l'autenticazione
   * @param {string} options.project - Progetto a cui appartiene il modulo
   * @param {string} options.module - Nome del modulo che genera i log
   * @param {string} [options.host='http://localhost:8081'] - Host del servizio
   * @param {number} [options.bufferSize=100] - Dimensione del buffer
   * @param {boolean} [options.autoFlush=true] - Se fare flush automaticamente
   * @param {number} [options.flushInterval=5000] - Intervallo di flush in ms
   * @param {number} [options.retryMaxAttempts=3] - Tentativi max in caso di errore
   * @param {number} [options.retryDelay=1000] - Ritardo tra tentativi in ms
   */
  constructor(options) {
    this.apiKey = options.apiKey;
    this.project = options.project;
    this.module = options.module;
    // Risolvi host da options, poi da BACKEND_URL o PRAMAIALOG_HOST (+ PRAMAIALOG_PORT), altrimenti fallback
    let resolvedHost = options.host || process.env.BACKEND_URL || process.env.PRAMAIALOG_HOST || 'http://localhost:8081';
    if (process.env.PRAMAIALOG_PORT && !/:\d+$/.test(resolvedHost)) {
      resolvedHost = `${resolvedHost.replace(/\/$/, '')}:${process.env.PRAMAIALOG_PORT}`;
    }
    this.host = resolvedHost.replace(/\/$/, '');
    this.bufferSize = options.bufferSize || 100;
    this.autoFlush = options.autoFlush !== false;
    this.flushInterval = options.flushInterval || 5000;
    this.retryMaxAttempts = options.retryMaxAttempts || 3;
    this.retryDelay = options.retryDelay || 1000;
    
    this.logBuffer = [];
    this.flushTimer = null;
    
    // Avvia il flush automatico se abilitato
    if (this.autoFlush) {
      this.startAutoFlush();
    }
  }
  
  /**
   * Avvia il flush automatico a intervalli regolari.
   * @private
   */
  startAutoFlush() {
    this.flushTimer = setInterval(() => {
      this.flush().catch(err => {
        console.error('Errore durante il flush automatico:', err);
      });
    }, this.flushInterval);
  }
  
  /**
   * Genera un UUID v4
   * @private
   * @returns {string} UUID v4
   */
  generateUUID() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
      const r = Math.random() * 16 | 0;
      const v = c === 'x' ? r : (r & 0x3 | 0x8);
      return v.toString(16);
    });
  }
  
  /**
   * Invia un log di livello DEBUG.
   * 
   * @param {string} message - Messaggio del log
   * @param {Object} [details=null] - Dettagli aggiuntivi
   * @param {Object} [context=null] - Contesto del log
   * @returns {string} ID del log
   */
  debug(message, details = null, context = null) {
    return this.log(LogLevel.DEBUG, message, details, context);
  }
  
  /**
   * Invia un log di livello INFO.
   * 
   * @param {string} message - Messaggio del log
   * @param {Object} [details=null] - Dettagli aggiuntivi
   * @param {Object} [context=null] - Contesto del log
   * @returns {string} ID del log
   */
  info(message, details = null, context = null) {
    return this.log(LogLevel.INFO, message, details, context);
  }
  
  /**
   * Invia un log di livello WARNING.
   * 
   * @param {string} message - Messaggio del log
   * @param {Object} [details=null] - Dettagli aggiuntivi
   * @param {Object} [context=null] - Contesto del log
   * @returns {string} ID del log
   */
  warning(message, details = null, context = null) {
    return this.log(LogLevel.WARNING, message, details, context);
  }
  
  /**
   * Invia un log di livello ERROR.
   * 
   * @param {string} message - Messaggio del log
   * @param {Object} [details=null] - Dettagli aggiuntivi
   * @param {Object} [context=null] - Contesto del log
   * @returns {string} ID del log
   */
  error(message, details = null, context = null) {
    return this.log(LogLevel.ERROR, message, details, context);
  }
  
  /**
   * Invia un log di livello CRITICAL.
   * 
   * @param {string} message - Messaggio del log
   * @param {Object} [details=null] - Dettagli aggiuntivi
   * @param {Object} [context=null] - Contesto del log
   * @returns {string} ID del log
   */
  critical(message, details = null, context = null) {
    return this.log(LogLevel.CRITICAL, message, details, context);
  }
  
  /**
   * Aggiunge un log al buffer.
   * 
   * @param {string} level - Livello del log
   * @param {string} message - Messaggio del log
   * @param {Object} [details=null] - Dettagli aggiuntivi
   * @param {Object} [context=null] - Contesto del log
   * @returns {string} ID del log
   */
  log(level, message, details = null, context = null) {
    const logId = this.generateUUID();
    const logEntry = {
      id: logId,
      timestamp: new Date().toISOString(),
      project: this.project,
      level: level,
      module: this.module,
      message: message,
      details: details,
      context: context
    };
    
    this.logBuffer.push(logEntry);
    
    // Se il buffer è pieno, fai un flush
    if (this.logBuffer.length >= this.bufferSize) {
      this.flush().catch(err => {
        console.error('Errore durante il flush del buffer pieno:', err);
      });
    }
    
    return logId;
  }
  
  /**
   * Invia tutti i log in coda al servizio.
   * 
   * @returns {Promise<boolean>} Promise che si risolve a true se l'invio è riuscito
   */
  async flush() {
    if (this.logBuffer.length === 0) {
      return true;
    }
    
    // Estrai tutti i log dal buffer
    const logs = [...this.logBuffer];
    this.logBuffer = [];
    
    // Invia i log al servizio
    const url = `${this.host}/api/logs/batch`;
    const headers = {
      'Content-Type': 'application/json',
      'X-API-Key': this.apiKey
    };
    
    // Fai più tentativi in caso di errore
    let attempt = 0;
    while (attempt < this.retryMaxAttempts) {
      try {
        const response = await fetch(url, {
          method: 'POST',
          headers: headers,
          body: JSON.stringify(logs)
        });
        
        if (response.status === 201) {
          return true;
        } else {
          console.error(`Errore nell'invio dei log: ${response.status} - ${await response.text()}`);
          attempt++;
          if (attempt < this.retryMaxAttempts) {
            await new Promise(resolve => setTimeout(resolve, this.retryDelay));
          }
        }
      } catch (error) {
        console.error(`Errore durante l'invio dei log: ${error.message}`);
        attempt++;
        if (attempt < this.retryMaxAttempts) {
          await new Promise(resolve => setTimeout(resolve, this.retryDelay));
        }
      }
    }
    
    // Se arriviamo qui, tutti i tentativi sono falliti
    // Riinserire i log nel buffer
    this.logBuffer = [...logs, ...this.logBuffer];
    
    return false;
  }
  
  /**
   * Chiude il logger e invia tutti i log rimanenti.
   * 
   * @returns {Promise<void>}
   */
  async close() {
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
      this.flushTimer = null;
    }
    
    await this.flush();
  }
}

/**
 * Funzione di utilità per configurare facilmente il logger
 * 
 * @param {Object} options - Opzioni di configurazione
 * @param {string} options.apiKey - API key per l'autenticazione
 * @param {string} options.project - Progetto PramaIA
 * @param {string} options.module - Nome del modulo
 * @param {string} [options.host] - Host del servizio di logging
 * @returns {PramaIALogger} Un'istanza configurata di PramaIALogger
 */
function setupLogger(options) {
  return new PramaIALogger(options);
}

// Esporta i tipi e le classi
module.exports = {
  LogLevel,
  LogProject,
  PramaIALogger,
  setupLogger
};

File: PramaIA-PDK/plugins/common/pramaialog.py
"""
Client Python per il servizio di logging PramaIA.

Questo modulo fornisce un client per inviare log al servizio
PramaIA-LogService. Può essere utilizzato da qualsiasi componente
Python dell'ecosistema PramaIA.

Esempio di utilizzo:
```python
from pramaialog import PramaIALogger, LogLevel, LogProject

# Crea un'istanza del logger
logger = PramaIALogger(
    api_key="your_api_key",
    project=LogProject.SERVER,
    module="workflow_service"
)

# Invia log di diversi livelli
logger.info("Servizio avviato")
logger.warning("Attenzione: file di configurazione non trovato", 
               details={"config_path": "/path/to/config"})
logger.error("Errore durante il caricamento del workflow", 
             details={"workflow_id": "123", "error": str(e)},
             context={"user_id": "admin"})
```
"""

import requests
import os
import logging
import json
import queue
import threading
import time
from datetime import datetime
from enum import Enum
from typing import Dict, Any, List, Optional, Union
import uuid

# Imposta il logger standard
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("pramaialog-client")

class LogLevel(str, Enum):
    """Livelli di log supportati."""
    DEBUG = "debug"
    INFO = "info"
    WARNING = "warning"
    ERROR = "error"
    CRITICAL = "critical"
    LIFECYCLE = "lifecycle"  # Livello speciale per tracciare il ciclo di vita dei documenti

class LogProject(str, Enum):
    """Progetti PramaIA supportati."""
    SERVER = "PramaIAServer"
    PDK = "PramaIA-PDK"
    AGENTS = "PramaIA-Agents"
    PLUGINS = "PramaIA-Plugins"
    OTHER = "other"

class PramaIALogger:
    """
    Client per il servizio di logging PramaIA.
    
    Fornisce metodi per inviare log di diversi livelli al servizio
    PramaIA-LogService.
    """
    
    def __init__(
        self,
        api_key: str,
        project: LogProject,
        module: str,
        host: Optional[str] = None,
        buffer_size: int = 100,
        auto_flush: bool = True,
        flush_interval: int = 5,
        retry_max_attempts: int = 3,
        retry_delay: int = 1
    ):
        """
        Inizializza il client di logging.
        
        Args:
            api_key: API key per l'autenticazione con il servizio
            project: Progetto a cui appartiene il modulo che genera i log
            module: Nome del modulo che genera i log
            host: Host del servizio di logging (incluso protocollo e porta)
            buffer_size: Dimensione massima del buffer prima del flush automatico
            auto_flush: Se True, invia automaticamente i log in batch
            flush_interval: Intervallo in secondi tra i flush automatici
            retry_max_attempts: Numero massimo di tentativi in caso di errore
            retry_delay: Ritardo in secondi tra i tentativi
        """
        self.api_key = api_key
        self.project = project
        self.module = module
        # Determina host dal parametro, oppure dalla variabile d'ambiente PRAMAIALOG_HOST,
        # o usa il fallback "http://localhost:8081"
        if host:
            resolved_host = host
        else:
            resolved_host = os.getenv('PRAMAIALOG_HOST') or os.getenv('BACKEND_URL') or 'http://localhost:8081'
        # Se PRAMAIALOG_PORT è impostata e resolved_host non contiene già una porta, aggiungila
        port_env = os.getenv('PRAMAIALOG_PORT')
        if port_env and ':' not in resolved_host.split('//')[-1]:
            resolved_host = f"{resolved_host.rstrip('/') }:{port_env}"
        self.host = resolved_host.rstrip("/")
        self.buffer_size = buffer_size
        self.auto_flush = auto_flush
        self.flush_interval = flush_interval
        self.retry_max_attempts = retry_max_attempts
        self.retry_delay = retry_delay
        
        # Buffer per i log
        self.log_buffer = queue.Queue(maxsize=buffer_size * 2)
        
        # Crea thread per flush automatico
        if auto_flush:
            self.running = True
            self.flush_thread = threading.Thread(target=self._auto_flush_worker)
            self.flush_thread.daemon = True
            self.flush_thread.start()
    
    def debug(
        self,
        message: str,
        details: Optional[Dict[str, Any]] = None,
        context: Optional[Dict[str, Any]] = None
    ) -> str:
        """Invia un log di livello DEBUG."""
        return self.log(LogLevel.DEBUG, message, details, context)
    
    def info(
        self,
        message: str,
        details: Optional[Dict[str, Any]] = None,
        context: Optional[Dict[str, Any]] = None
    ) -> str:
        """Invia un log di livello INFO."""
        return self.log(LogLevel.INFO, message, details, context)
    
    def warning(
        self,
        message: str,
        details: Optional[Dict[str, Any]] = None,
        context: Optional[Dict[str, Any]] = None
    ) -> str:
        """Invia un log di livello WARNING."""
        return self.log(LogLevel.WARNING, message, details, context)
    
    def error(
        self,
        message: str,
        details: Optional[Dict[str, Any]] = None,
        context: Optional[Dict[str, Any]] = None
    ) -> str:
        """Invia un log di livello ERROR."""
        return self.log(LogLevel.ERROR, message, details, context)
    
    def critical(
        self,
        message: str,
        details: Optional[Dict[str, Any]] = None,
        context: Optional[Dict[str, Any]] = None
    ) -> str:
        """Invia un log di livello CRITICAL."""
        return self.log(LogLevel.CRITICAL, message, details, context)
        
    def lifecycle(
        self,
        message: str,
        details: Optional[Dict[str, Any]] = None,
        context: Optional[Dict[str, Any]] = None
    ) -> str:
        """
        Invia un log di livello LIFECYCLE.
        
        Questo livello è specifico per tracciare gli eventi del ciclo di vita dei documenti
        come creazione, modifica, cancellazione, upload, ecc.
        
        Args:
            message: Il messaggio da registrare
            details: Dizionario con dettagli aggiuntivi
            context: Contesto dell'evento
            
        Returns:
            ID del log creato
        """
        # Aggiungi automaticamente un marker al messaggio
        marked_message = f"🔄 [LIFECYCLE] {message}"
        
        # Aggiungi il tag log_type per garantire la compatibilità con le versioni precedenti
        lifecycle_details = details.copy() if details else {}
        lifecycle_details["log_type"] = "lifecycle"
            
        # Se non c'è già un lifecycle_event, aggiungiamolo
        if "lifecycle_event" not in lifecycle_details:
            # Estrai un nome di evento dal messaggio o usa un valore predefinito
            import re
            event_match = re.search(r'\b([A-Z_]+)\b', message)
            lifecycle_details["lifecycle_event"] = event_match.group(1) if event_match else "GENERIC_EVENT"
        
        # Usa il livello LIFECYCLE direttamente
        return self.log(LogLevel.LIFECYCLE, marked_message, lifecycle_details, context)
    
    def log(
        self,
        level: LogLevel,
        message: str,
        details: Optional[Dict[str, Any]] = None,
        context: Optional[Dict[str, Any]] = None
    ) -> str:
        """
        Aggiunge un log al buffer.
        
        Args:
            level: Livello del log
            message: Messaggio del log
            details: Dettagli aggiuntivi (opzionale)
            context: Contesto del log (opzionale)
            
        Returns:
            ID del log creato
        """
        log_id = str(uuid.uuid4())
        
        # Assicurati che i dettagli del lifecycle abbiano il tag appropriato
        if level == LogLevel.LIFECYCLE:
            if not details:
                details = {}
            if isinstance(details, dict) and "log_type" not in details:
                details["log_type"] = "lifecycle"
        
        log_entry = {
            "id": log_id,
            "timestamp": datetime.now().isoformat(),
            "project": self.project,
            "level": level,  # Usa il livello originale, non convertire LIFECYCLE in INFO!
            "module": self.module,
            "message": message,
            "details": details,
            "context": context
        }
        
        try:
            self.log_buffer.put(log_entry, block=False)
            
            # Se il buffer è pieno, fai un flush
            if self.log_buffer.qsize() >= self.buffer_size:
                self.flush()
        except queue.Full:
            # Buffer pieno, fai un flush e poi aggiungi
            logger.warning("Buffer dei log pieno, flush forzato")
            self.flush()
            self.log_buffer.put(log_entry)
        
        return log_id
    
    def flush(self) -> bool:
        """
        Invia tutti i log in coda al servizio.
        
        Returns:
            True se tutti i log sono stati inviati con successo, False altrimenti
        """
        if self.log_buffer.empty():
            return True
        
        # Estrai tutti i log dal buffer
        logs = []
        try:
            while not self.log_buffer.empty():
                logs.append(self.log_buffer.get(block=False))
                self.log_buffer.task_done()
        except queue.Empty:
            pass
        
        if not logs:
            return True
        
        # Invia i log al servizio
        url = f"{self.host}/api/logs/batch"
        headers = {
            "Content-Type": "application/json",
            "X-API-Key": self.api_key
        }
        
        # Fai più tentativi in caso di errore
        attempt = 0
        while attempt < self.retry_max_attempts:
            try:
                response = requests.post(url, headers=headers, json=logs, timeout=10)
                
                if response.status_code == 201:
                    return True
                else:
                    logger.error(f"Errore nell'invio dei log: {response.status_code} - {response.text}")
                    attempt += 1
                    if attempt < self.retry_max_attempts:
                        time.sleep(self.retry_delay)
            except Exception as e:
                logger.error(f"Errore durante l'invio dei log: {str(e)}")
                attempt += 1
                if attempt < self.retry_max_attempts:
                    time.sleep(self.retry_delay)
        
        # Se arriviamo qui, tutti i tentativi sono falliti
        # Riinserire i log nel buffer
        for log in logs:
            try:
                self.log_buffer.put(log, block=False)
            except queue.Full:
                logger.error(f"Impossibile riaggiungere un log al buffer: {log}")
        
        return False
    
    def _auto_flush_worker(self):
        """Thread worker per il flush automatico."""
        while self.running:
            time.sleep(self.flush_interval)
            try:
                self.flush()
            except Exception as e:
                logger.error(f"Errore durante il flush automatico: {str(e)}")
    
    def close(self):
        """
        Chiude il logger, flushando tutti i log in coda.
        """
        self.running = False
        if hasattr(self, 'flush_thread') and self.flush_thread.is_alive():
            self.flush_thread.join(timeout=2)
        
        self.flush()


# Funzione di utilità per configurare facilmente il logger
def setup_logger(
    api_key: str,
    project: Union[LogProject, str],
    module: str,
    host: Optional[str] = None
) -> PramaIALogger:
    """
    Configura e restituisce un'istanza di PramaIALogger.
    
    Args:
        api_key: API key per l'autenticazione
        project: Progetto PramaIA (LogProject o stringa)
        module: Nome del modulo
        host: Host del servizio di logging
        
    Returns:
        Un'istanza configurata di PramaIALogger
    """
    # Converti la stringa in enum se necessario
    if isinstance(project, str):
        try:
            project = LogProject(project)
        except ValueError:
            project = LogProject.OTHER
    
    return PramaIALogger(
        api_key=api_key,
        project=project,
        module=module,
        host=host
    )
